Source/Services.Clients/ReverseCallClient.cs

```diff
@@ -244,121 +159,293 @@ protected virtual void Dispose(bool disposing)
+            try
+            {
+                var responseContext = new ReverseCallResponseContext { CallId = callId.ToProtobuf() };
+                _protocol.SetResponseContext(responseContext, response);
+                var message = new TClientMessage();
+                _protocol.SetResponse(response, message);
+
+                await WriteMessage(message, cancellationToken).ConfigureAwait(false);
+            }
+            catch (Exception exception)
+            {
+                _metrics.IncrementTotalFailedResponseWrites();
+                _logger.HandlingRequestFailedToWriteResponse(typeof(TRequest), callId, exception);
+            }
+        }
+
+        async Task WriteMessage(TClientMessage message, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            _logger.WritingMessage(typeof(TClientMessage));
+
+            if (!_writeLock.Wait(0))
             {
-                _logger.WritingResponse(callId);
-                return _clientToServer.WriteAsync(message);
+                var stopwatch = Stopwatch.StartNew();
+
+                _metrics.IncrementPendingWrites();
+                _logger.WritingMessageBlockedByAnotherWrite(typeof(TClientMessage));
+
+                await _writeLock.WaitAsync(cancellationToken).ConfigureAwait(false);
+
+                try
+                {
+                    stopwatch.Stop();
+                    _metrics.DecrementPendingWrites();
+                    _metrics.AddToTotalWriteWaitTime(stopwatch.Elapsed);
+                    _logger.WritingMessageUnblockedAfter(typeof(TServerMessage), stopwatch.Elapsed);
+                }
+                catch
+                {
+                }
             }
 
-            _logger.ClientCancelled(callId);
-            return Task.CompletedTask;
+            try
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var stopwatch = Stopwatch.StartNew();
+
+                await _clientToServer.WriteAsync(message).ConfigureAwait(false);
+
+                stopwatch.Stop();
+                var messageSize = message.CalculateSize();
+                _metrics.AddToTotalWriteTime(stopwatch.Elapsed);
+                _metrics.IncrementTotalWrites();
+                _metrics.AddToTotalWriteBytes(messageSize);
+                _logger.WroteMessage(typeof(TClientMessage), stopwatch.Elapsed, messageSize);
+            }
+            finally
+            {
+                _writeLock.Release();
+            }
         }
 
         void ThrowIfInvalidPingInterval(TimeSpan pingInterval)
         {
             if (pingInterval.TotalMilliseconds <= 0) throw new PingIntervalNotGreaterThanZero();
         }
 
-        void ThrowIfConnecting()
+        void EnsureOnlyConnectingOnce()
         {
-            if (_connecting)
+
+            if (_connecting) throw new ReverseCallClientAlreadyCalledConnect();
+            lock (this)
```

Comment:
I don't know much about locks in C#, but I've heard that using `lock(this)` can be [dangerous](https://stackoverflow.com/a/251668/5806412). [MS docs](https://stackoverflow.com/a/251668/5806412) also advice against it.